### ISSUE TITLE
promise: spawn_into_new_thread can park its future forever (lost wakeup; panic in the closure)

### ISSUE THREAD

**Issue by PHPCraftdream** [wezterm/wezterm#8062]
**What Operating System(s) are you seeing this problem on?**

Windows. The defects are in `promise`, which is platform-independent; the
reproductions are unit tests and should behave the same anywhere.

**WezTerm version**

Source at `fe3006aef` (main as of 2026-08-17). Not tied to a release: the code
in question has been in `promise/src/spawn.rs` since the crate was written.

**Did you try the latest nightly build?**

The reproductions run against current main, so yes in the sense that matters
here. They are crate-level tests rather than a GUI scenario, so a nightly
binary would not tell you anything the tests do not.

**Configuration**

No config. Nothing here is configuration-dependent.

---

## Describe the bug

`promise::spawn::spawn_into_new_thread` can leave the future it returns parked
forever, with the result already sitting in the channel and no poll ever
scheduled to collect it. Two independent paths get there.

**(A) Lost wakeup.** `poll` does one `try_recv`, and only afterwards stores the
waker:

```rust
match self.rx.try_recv() {
    Ok(res) => Poll::Ready(res),
    Err(TryRecvError::Empty) => {
        let mut waker = self.holder.waker.lock().unwrap();
        waker.replace(cx.waker().clone());
        Poll::Pending
    }
    ...
}
```

A worker that sends in between takes the waker mutex, finds `None`, wakes
nobody, and exits. `poll` then stores a waker nothing will ever use and returns
`Pending`. The result is in the channel the whole time; one more poll would
deliver it, and no poll is coming.

**(B) A panic in `f` unwinds past both the send and the wake.** `poll` already
knows how to report a worker that produced no result -- it turns the
disconnected channel into an error -- but it only gets the chance if something
wakes it up to look. Nothing does.

Effect on the GUI, for (B) specifically: `start_overlay` runs

```rust
let res = func(tab_id, tw_term);
TermWindow::schedule_cancel_overlay(window, tab_id, Some(overlay_pane_id));
res
```

so a panic in `func` skips `schedule_cancel_overlay`, and the overlay pane
stays on screen with nothing left to close it. The leaked task is the quiet
part; the stuck overlay is the part a user would report.

Both are reachable from every current caller of `spawn_into_new_thread`:
`wezterm-gui/src/overlay/mod.rs:48` and `:84`, `mux/src/termwiztermtab.rs:572`,
`wezterm-client/src/domain.rs:951`.

## To Reproduce

Branch: https://github.com/PHPCraftdream/wezterm/tree/repro-8060 (based on
`fe3006aef`). It carries upstream's unfixed `promise/src/spawn.rs` plus four
tests, and the fix as a separate patch rather than as code.

```
git clone -b repro-8060 https://github.com/PHPCraftdream/wezterm.git
cd wezterm
bash repro-8060/verify.sh
```

That runs the whole before/after matrix and prints:

```
BEFORE -- upstream code as it stands (fe3006aef)
  control: happy path      PASS  as expected
  defect (2): lost wakeup  HANG  as expected  killed after 30s
  defect (1): worker panic FAIL  as expected  called `Result::unwrap()` on an `Err` value: "SendError(..)"
  defect (3): panic in f() HANG  as expected  killed after 30s

AFTER -- same file plus the production fix from PR #8060
  control: happy path      PASS
  defect (2): lost wakeup  PASS
  defect (1): worker panic PASS
  defect (3): panic in f() PASS
```

`repro-8060/README.md` has the same thing step by step if you would rather
watch each one.

The two windows are a handful of instructions wide, so racing threads and
hoping does not hit them. Two `cfg(test)` hooks -- one in `poll`, one in the
worker thread body -- turn the window into a rendezvous: `poll` announces that
it has entered it and waits there until the worker has finished completely.
They change nothing about *what* either side does, only *when*, and they are
identical either side of the patch, so exactly one thing moves between "before"
and "after".

## Expected Behavior

A result that was successfully sent is delivered to the awaiting future, and a
worker that dies without producing one is reported as an error rather than
leaving the caller parked.

## A third defect, which I am reporting as hardening rather than a bug

`tx.send(res).unwrap()` panics the worker thread if the receiving future has
been dropped -- a cancelled caller takes a thread down with it. Nothing reads
the result at that point, so discarding it is the correct response.

I said on #8060 that this looked reachable via `attach()` in
`wezterm-client/src/domain.rs` and that I had not traced it. I have now, and it
does not hold up. All four call sites either `.detach()` the task or are driven
by one that does; there is no `select`/`Timer`/race cancelling any of them; and
neither production user of `ScopedExecutor` (`wezterm/src/cli/mod.rs:206`,
`wezterm-gui/src/main.rs:553`) reaches `spawn_into_new_thread`. The only way to
drop that future today is process teardown, where the panic is invisible.

So the panic is real and the test demonstrates it, but nothing in the tree
reaches it. Treat it as a trap for the next caller, not a live defect.

## Anything else?

PR #8060 fixes all three. It predates this issue -- you asked for the bug
report separately so you could judge the problem before the implementation, and
that is the right order, so the description above deliberately does not argue
for any particular fix.

One thing that does need your call rather than mine: the rendezvous hooks are
what make (A) and (B) deterministic, and they live in production source under
`cfg(test)`. They are not in #8060, which ships the fix with tests that do not
need them. If you would rather have the deterministic versions in-tree, that
means carrying the hooks; if not, the branch above stays a scratch artifact.
